notebooks/02-ingestion/10-updating-data.ipynb

```diff
@@ -0,0 +1,1188 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "26b2aff3-b485-42ac-868a-a75ce893deb2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The smallest slice of a time chunk that can be updated is 1 second\n",
+    "# Druid's segment overlapping and versioning strategy requires that the new segment include\n",
+    "# all of the rows in the time interval where a row is going to be changed. \n",
+    "# This SQL shows what other rows occur in the same second? \n",
+    "sql='''\n",
+    "SELECT * FROM \"example-flights-updates\" \n",
+    "WHERE TIME_IN_INTERVAL(\"__time\", '2005-11-01T00:15:00.000Z/PT1S') \n",
+    "'''\n",
+    "display.sql(sql)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "29c99123-3092-4102-b3f4-20bb804227cb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#before updating anything, get a count and a checksum so we can validated data consistency\n",
+    "sql='''\n",
+    "SELECT count(*) \"total_flights_day\", SUM(\"departure_delay\") \"total_time_lost\" \n",
+    "FROM \"example-flights-updates\"\n",
+    "WHERE TIME_IN_INTERVAL(\"__time\", '2005-11-01T00:00:00.000Z/P1D')  -- totals for the day\n",
+    "'''\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "97796d23-da51-4aca-ae28-5a8fbedb7ed0",
+   "metadata": {},
+   "source": [
+    "For this update, given that it is a single row, you can get away with creating a segment that only covers 1 second of time.\n",
+    "Like you can see above, besides the record that needs updating, there is only one other record during that second that we will need to include. \n",
+    "\n",
+    "A few important points on how to use this method:\n",
+    "\n",
+    "- Use a SQL REPLACE statement that overwrites the minimum timeframe where this update occurs. That's `00:15:00 <= t < 00:15:01` of `2005-11-01` of the timeline as expressed in the OVERWRITE WHERE clause in the SQL below. \n",
+    "- The SELECT portion of the request, must read the data for the target timeframe, including the row that is being updated and any other rows that exist in the same timeframe. In the SQL below it is: `WHERE TIME_IN_INTERVAL(\"__time\", '2005-11-01T00:15:00.000Z/PT1S')`\n",
+    "- The `PARTITIONED BY` clause must fit exactly within the overwrite timeframe. The SQL below uses `FLOOR(__time TO SECOND)` which creates second level partitions.\n",
+    "- The update columns can be expressed as conditional CASE expressions that change the value of the column when the key columns match the target update (rows). The expression used in the SQL below is:\n",
+    "  \n",
+    "  ```\n",
+    "      CASE \n",
```

Comment:
Since we are updating a single row, the WHERE clause is actually in the CASE WHEN clause.. It only changes the value of the column for the row we want to change. All other rows in the 1 second interval are left "as is".